cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -792,60 +818,81 @@ namespace cooperative_lanechange
-    {
-        lanelet::BasicPoint2d veh_point(state.X_pos_global, state.Y_pos_global);
-                double min_distance = std::numeric_limits<double>::max();
-        int i = 0;
+    int CooperativeLaneChangePlugin::get_nearest_route_point_index(lanelet::BasicLineString2d& points, double ending_downtrack){
         int best_index = 0;
-        for (const auto& p : points)
-        {
-            double distance = lanelet::geometry::distance2d(p,veh_point);
-            if (distance < min_distance)
-            {
-            best_index = i;
-            min_distance = distance;
+        bool set_index = false;
+        for(int i=0;i < points.size();i++){
+            double downtrack = wm_->routeTrackPos(points[i]).downtrack;
+            if(downtrack > ending_downtrack){
+                best_index = i - 1;
+                ROS_DEBUG_STREAM("get_nearest_route_point_index>> Found best_idx: " << best_index<<", points[i].x(): " << points[best_index].x() << ", points[i].y(): " << points[best_index].y() << ", downtrack: "<< downtrack);
+                set_index = true;
+                break;
             }
-            i++;
         }
-        return best_index;
+        if (!set_index)
+        {
+            best_index = points.size() - 1;
+        }
+        ROS_DEBUG_STREAM("get_nearest_route_point_index>> Found best_idx: " << best_index<<", points[i].x(): " << points[best_index].x() << ", points[i].y(): " << points[best_index].y());
 
+        return best_index;
     }
 
-        int CooperativeLaneChangePlugin::get_ending_point_index(lanelet::BasicLineString2d& points, double ending_downtrack){
-        int best_index = points.size()-1;
-        for(int i=points.size()-1;i>=0;i--){
+    int CooperativeLaneChangePlugin::get_nearest_point_index(const std::vector<lanelet::BasicPoint2d>& points,
+                                               const cav_msgs::VehicleState& state) const
+    {
+        lanelet::BasicPoint2d state_pos(state.X_pos_global, state.Y_pos_global);
+        double ending_downtrack = wm_->routeTrackPos(state_pos).downtrack;
+        ROS_DEBUG_STREAM("get_nearest_point_index: state_pos: " << state_pos.x() << ", " << state_pos.y() << ", ending_downtrack" << ending_downtrack);
+        int best_index = 0;
+        bool set_index = false;
+        for(int i=0;i < points.size();i++){
             double downtrack = wm_->routeTrackPos(points[i]).downtrack;
-            if(downtrack <= ending_downtrack){
-                best_index = i;
+
+            if(downtrack > ending_downtrack){
+                best_index = i - 1;
+                ROS_DEBUG_STREAM("get_nearest_point_index>> Found best_idx: " << best_index<<", points[i].x(): " << points[best_index].x() << ", points[i].y(): " << points[best_index].y() << ", downtrack: "<< downtrack);
+                set_index = true;
                 break;
             }
         }
+        if (!set_index)
```

Comment:
Same as above. best_index can be initialized with points.size() - 1, to remove the need for this check